libs/keyboard/keyboard.ts

```diff
@@ -3,5 +3,154 @@
  */
 //% icon="\uf11c" color="#303030"
 namespace keyboard {
+    // prevent catastrophic pounding of keyboard
+    class State {
+        private _keysDown: number[];
+        private _mediasDown: number[];
+        private _functionsDown: number[];
+        constructor() { }
 
+        type(text: string) {
+            if (!text) return;
+
+            if (this._keysDown) {
+                for (let i = 0; i < text.length; ++i) {
+                    const c = text.charCodeAt(i);
+                    const ic = this._keysDown.indexOf(c);
+                    if (ic > -1)
+                        this._keysDown.splice(ic, 1);
+                }
+            }
+            __type(text);
+        }
+
+        key(key: string, event: KeyboardKeyEvent) {
+            if (!key) return;
+
+            if (!this._keysDown) this._keysDown = [];
+            const c = key.charCodeAt(0);
+            if (State.updateState(this._keysDown, c, event))
+                __key(c, event);
+        }
+
+        mediaKey(key: KeyboardMediaKey, event: KeyboardKeyEvent) {
+            if (!this._mediasDown) this._mediasDown = [];
+            if (State.updateState(this._mediasDown, key, event))
+                __mediaKey(key, event);
+        }
+
+        functionKey(key: KeyboardFunctionKey, event: KeyboardKeyEvent) {
+            if (!this._functionsDown) this._functionsDown = [];
+            if (State.updateState(this._functionsDown, key, event))
+                __functionKey(key, event);
+        }
+
+        private static updateState(downKeys: number[], c: number, event: KeyboardKeyEvent): boolean {
+            let i = downKeys.indexOf(c);
+            switch (event) {
+                // clear down
+                case KeyboardKeyEvent.Press:
+                    if (i > -1)
+                        downKeys.splice(i, 1);
+                    return true;
+                // must be down
+                case KeyboardKeyEvent.Up:
+                    if (i > -1) {
+                        downKeys.splice(i, 1);
+                        return true;
+                    } else {
+                        return false;
+                    }
+                // must be down
+                case KeyboardKeyEvent.Down:
+                    if (i > -1) {
+                        return false;
+                    } else {
+                        downKeys.push(c)
+                        return true;
+                    }
+            }
+
+            return false;
+        }
+
+        clear() {
+            // send up command for all down keys and clear
+            if (this._keysDown) {
+                this._keysDown.forEach(c => __key(c, KeyboardKeyEvent.Up));
```

Comment:
How does for each interact with modifying the array?

Reply:
using pop now